### DISCUSSION TITLE
How to include a code snippet in a GitHub issue comment?

### DISCUSSION THREAD

**Discussion by UlyssesZh** [community/community#61499] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

According to [this doc](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet#linking-to-code), just pasting the permalink of the range of codes would be fine. However, when I tried that in [this issue](https://github.com/audiojs/audio-decode/issues/34#issue-1816556692), the link was shown as is instead of showing a fancy code snippet. What did I do wrong?

**Comment by Matt-Osborn** [comment]
in general you can use back ticks  (they are on the same key as the tilde ~ on most qwerty keyboards) - this works anywhere that renders markup. example 

`print("hello world")`

also, for code blocks you can use three:

```
def function():
     print("this is an example")
```

**Reply by UlyssesZh** [reply]
You misunderstood me. This is the look I want to have:
![rendered code snippet](https://docs.github.com/assets/cb-68457/mw-1440/images/help/repository/rendered-code-snippet.webp)

**Comment by Matt-Osborn** [comment]
Yes I did misunderstand - and your right, it makes it sound like it should work in the same repo - which it looks like you are.

Maybe it works on 'comments' but not issues...

**Reply by UlyssesZh** [reply]
Look at [my case](https://github.com/audiojs/audio-decode/issues/34#issue-1816556692). It **is** in the repository it originates in.

**Reply by Noahnoah55** [reply, marked as answer]
Just had a similar problem. You gotta remove the "C2" part from each line number. Also since you're just doing one line you might also want to just use L23 instead of L23-L23

Instead of `https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23C2-L23C2` you want `https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23`

**Reply by jaamarks** [reply]
> Just had a similar problem. You gotta remove the "C2" part from each line number. Also since you're just doing one line you might also want to just use L23 instead of L23-L23
> 
> Instead of `https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23C2-L23C2` you want `https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23`

This was helpful. Also, I learned that this extra `CX` portion is included when you try to drag your mouse to highlight the code. But if you click **the number to the left** of the code. Hold down shift. **Click the actual number of the last line** of code you want to highlight. Then copy permalink. This will give you a URL that will show the code you want when pasted in an issue.

**Comment by jcubic** [comment]
It looks like documented feature that doesn't work at all.

Or just the screenshot is misleading, something like this was working, I've seen code snippets in issues.

**Comment by cipher20000** [comment]
To include a code snippet in a GitHub issue comment:

- For inline code, wrap your code with single backticks, like `print("hello world")`.
- For multi-line/code blocks, use triple backticks (```) before and after your code:

```
def my_function():
    print("example")
```

- For linking to a specific line/range in a file, copy the permalink and use the format ending in #Lnumber (like #L23). Remove 'C2' or similar column reference from the URL (only use #Lnumber or #Lstart-Lend). Sometimes, only the link will show.
- Screenshots are a useful alternative if the snippet doesn't display as expected.

Right now, GitHub sometimes does not render permalinks as embedded code in comments. In that case, use the code formatting methods above or attach an image. Hope this helps!

**Reply by jcubic** [reply]
The question is how to embed GitHub source code by adding link to the line range. It's part of GitHub documentation, but it doesn't work.

**Reply by monsterhxw** [reply]
<img width="583" height="514" alt="image" src="https://github.com/user-attachments/assets/301cf70d-155d-4a79-ae80-babfac5286ae" />

See: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet#linking-to-code

**Reply by jcubic** [reply]
@monsterhxw yes, this is the documentation that doesn't work.

**Reply by HalfWhitt** [reply]
@jcubic I just came across this issue, because I was trying to link to a different repo and wasn't sure why it wasn't working. However, it does work fine for me when linking within the same repo. Perhaps you could share a literal example of a permalink you've tried pasting, and had it not work?

Edit: I tested out the URL fix [suggested above](https://github.com/orgs/community/discussions/61499#discussioncomment-9847147) (`https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23`) on the original issue linked to, and [it works for me](https://github.com/audiojs/audio-decode/issues/34#issue-1816556692).

Screenshot:
<img width="836" height="95" alt="screenshot of code embed" src="https://github.com/user-attachments/assets/c37ea008-c61b-4070-b968-ac6bbfea294b" />

You can confirm whether it works for you or not by pasting it as a reply there and clicking "Preview", without necessarily posting it.

**Comment by jpstotz** [comment]
There is another blocker for displaying the code lines: if you have enabled git blame in the code view and then copy the link to one or more lines and paste it into an issue you don't get the code vierw. This is because the link doesn't start like 

`https://github.com/audiojs/audio-decode/blob/...` 

but

`https://github.com/audiojs/audio-decode/blame/...`

If you manually change the `blame` back to `blob` you get the code view.

**Comment by prashantkoirala465** [comment]
On GitHub, simply pasting a permalink to a line range does not automatically render a fancy embedded code snippet in issue comments.

The “rich code preview” (the boxed snippet view) only works in certain places like pull request descriptions, PR review comments, or when using the “Insert a suggestion” feature. Regular issue comments don’t auto-expand permalinks into code previews.

So you didn’t necessarily do anything wrong it’s just a limitation of where GitHub supports snippet embedding.

If you want to include code in an issue comment, the reliable way is to use Markdown code fences:

<pre> ```language // your code here ``` </pre>

Example:

<pre> ```cpp int add(int a, int b) { return a + b; } ``` </pre>

If you want to reference a specific line in a file, you can still paste the permalink, but it will stay as a normal link.

**Reply by HalfWhitt** [reply]
In my experience, it does work in [comments on issues](https://github.com/audiojs/audio-decode/issues/34#issue-1816556692), as long as it's in the same repo. Does it not render for you? Does it vary between repos somehow?

**Comment by RAAZIA-YASIN6770** [comment]
To include a code snippet, you can use triple backticks (```) before and after your code. For example:
```javascript
console.log("Hello World");
```
This will format it properly with syntax highlighting.

**Reply by RAAZIA-YASIN6770** [reply]
detailed:
You can include a code snippet in a GitHub issue or discussion by using triple backticks (```).

1. For Inline Code:
Use a single backtick before and after the text: `your-code-here`

2. For a Code Block:
Use three backticks before and after your code. You can also add the language name (like python, javascript, or html) after the first set of backticks for syntax highlighting:

```python
def hello_world():
print("Hello, GitHub!")
```

3. Pro-Tip:
You can also use the "Insert code" icon (< >) in the formatting toolbar above the comment box to automatically insert these backticks.

If this helps you, please consider marking this as the answer! 🚀

**Reply by HalfWhitt** [reply]
The question is about displaying a link to GitHub code as a snippet or "preview" of that code, not just code-formatting text in the post.

**Comment by Shreyas-S-809** [comment]
GitHub uses Markdown, so adding code is just formatting — no special buttons needed.

---

## Inline code (for small things)

Use a single backtick ` `

Example:

`npm install`

It shows up like: `npm install`

Good for commands, file names, variables, flags, etc.

---

## Code block (the one you usually want)

Use **three backticks** before and after the code.

<pre>
```
your code here
```
</pre>

---

## With syntax highlighting (much nicer)

Add the language name right after the first three backticks.

<pre>
```python
def hello():
    print("Hello world")
```
</pre>

GitHub will color it automatically.

Common ones:

* `js` / `javascript`
* `ts`
* `python`
* `bash` / `sh`
* `json`
* `yaml`
* `cpp`
* `java`

---

## Pro tip (people miss this)

If your code itself contains backticks, wrap it with **four backticks** outside:

<pre>
````
```
actual code that has backticks
```
````
</pre>

---

That’s it — GitHub comments, issues, PRs, discussions all follow the same rules.

**Reply by HalfWhitt** [reply]
That's not what this question is about. The four backticks thing is good to know, though.

**Comment by Dhilipkumar-max** [comment]
Hi,

When you paste a permalink to a specific range of code, GitHub only renders it as a rich code snippet preview in certain contexts. In many cases (including some issues and discussions), the link will remain a normal URL instead of expanding automatically.

**A few things to check:**

Make sure the link is a permalink to a specific commit, not to a branch (e.g., main). You can generate a true permalink by pressing y while viewing the file on GitHub.

Ensure the link includes the line range (for example: #L10-L25).

The preview only works when the link is pasted on its own line (not embedded inside text or backticks).

If all of the above are correct and it still doesn’t expand, it may simply be a limitation of the context where you’re posting. GitHub does not always render rich previews inside issues or certain discussion views.

**Reply by david-g-alayachew-usps** [reply]
> Make sure the link is a permalink to a specific commit, not to a branch (e.g., main).

This is the missing detail folks don't seem to understand. This feature only works if you point to a **_commit_**, not a _branch_ or a _tag_.

Even look at what happens when you highlight some code, click the triple-dots, then click "Copy Permalink" -- it will give you a link to a _**commit**_, even if you were in a branch or tag.

**Reply by david-g-alayachew-usps** [reply]
This answer should have been the accepted one.

**Reply by david-g-alayachew-usps** [reply]
Maybe edit your answer to include about the triple-dots --> Copy Permalink? Would help folks in the future, as opposed to them trying to piece it together from this disorganized thread of answers.

**Comment by alliecatowo** [comment]
Was having this issue on a "request changes" comment, couldn't figure it out.

For me, it needed the full long form SHA, not abbreviated.

For example:
https://github.com/community/community/blob/7d09970/README.md?plain=1#L1-L3 (short SHA)

https://github.com/community/community/blob/7d09970c15877cd8fb404dc7b8aa27cfc542f698/README.md?plain=1#L1-L3

^^ long SHA.

Might not be the issue others are having, but I didn't see this well documented anywhere, so it's worth noting. Especially if you are using Claude to write comments lol

**Comment by yugenelee** [comment]
There are two different ways to include code in a GitHub comment:

1. Embed code from a GitHub repository using a permalink
2. Paste the code directly into a fenced code block

For the rich embedded snippet shown in GitHub's documentation, the URL must meet several conditions.

## 1. Use a permanent link to a commit

The URL should point to a specific commit, not a moving branch such as `main`:

```text
https://github.com/OWNER/REPOSITORY/blob/FULL_COMMIT_SHA/path/to/file#L10-L20
```

The safest way to create it is:

1. Open the file on GitHub.
2. Click the number beside the first line.
3. Hold `Shift` and click the final line number.
4. Open the menu beside the selected lines.
5. Select **Copy permalink**.

This follows GitHub's official instructions for [creating a permanent link to a code snippet](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet).

Using **Copy permalink** is preferable to constructing the URL manually because it points to a specific version of the file using a commit SHA.

## 2. Remove column positions from the fragment

A URL copied after selecting text inside a line may end with column positions such as:

```text
#L23C2-L23C18
```

That creates a valid link to a precise text range, but in this case it prevents GitHub from recognising the URL as an embeddable line permalink.

Use line anchors only:

```text
#L23
```

or:

```text
#L23-L30
```

In the original example, this:

```text
https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23C2-L23C2
```

should become:

```text
https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23
```

## 3. Make sure the URL uses `blob`, not `blame`

If the link was copied while Git blame was open, it may contain:

```text
https://github.com/OWNER/REPOSITORY/blame/COMMIT_SHA/path/to/file#L10-L20
```

Change `blame` to `blob`, or reopen the normal file view and use **Copy permalink** again:

```text
https://github.com/OWNER/REPOSITORY/blob/COMMIT_SHA/path/to/file#L10-L20
```

## 4. Paste the URL on its own line

Do not place the permalink inside Markdown link syntax, backticks or a sentence. Paste the complete URL on a separate line:

```text
Some explanation before the snippet.

https://github.com/OWNER/REPOSITORY/blob/FULL_COMMIT_SHA/path/to/file#L10-L20

Some explanation after the snippet.
```

GitHub can then recognise the URL and replace it with the embedded code view.

## 5. The comment must be in the same repository

According to GitHub's documentation, the rich code snippet is rendered only when the comment belongs to the repository from which the permalink originated.

If you paste the link into another repository, GitHub displays a normal URL instead. This feature works in comments, not inside committed Markdown files such as `README.md`.

To link to specific lines in a Markdown file, open its plain source using `?plain=1` before the line fragment:

```text
https://github.com/OWNER/REPOSITORY/blob/FULL_COMMIT_SHA/README.md?plain=1#L10-L20
```

## Reliable fallback

If the source comes from another repository, is not committed yet, or still refuses to expand, paste it directly using a fenced code block:

````markdown
```javascript
function hello() {
  console.log("Hello");
}
```
````

Adding the language name after the opening fence enables syntax highlighting. GitHub documents this under [creating and highlighting code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks), and this [fenced code block reference](https://markdific.com/docs/markdown/fenced-code-blocks/) covers the syntax, language identifiers and common formatting mistakes in more detail.

For the original case, the important fix is to remove the `C2` column positions and use:

```text
https://github.com/audiojs/audio-decode/blob/83c3883f5a2505de50da880c7c58a516d4415793/package.json#L23
```

Because it points to a specific commit, uses `blob`, contains only a line anchor and is being posted in the same repository, GitHub should render it as an embedded code snippet.

**Reply by jcubic** [reply]
This question has a correct answer. You don't need to add a duplicated one.

Also note about:

https://dontpastetheai.com
